rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+            .map(|upload| (MultipartId::default(), Box::new(upload) as _))
+    }
+
+    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> OSResult<()> {
+        MultiPartStore::abort_multipart(self.0.as_ref(), location, multipart_id).await
+    }
+
+    async fn get_opts(&self, location: &Path, options: GetOptions) -> OSResult<GetResult> {
+        self.0.get_opts(location, options).await
+    }
+
+    async fn get_range(&self, location: &Path, range: Range<usize>) -> OSResult<Bytes> {
+        self.0.get_range(location, range).await
+    }
+
+    async fn get_ranges(&self, location: &Path, ranges: &[Range<usize>]) -> OSResult<Vec<Bytes>> {
+        self.0.get_ranges(location, ranges).await
+    }
+
+    async fn head(&self, location: &Path) -> OSResult<ObjectMeta> {
+        self.0.head(location).await
+    }
+
+    async fn delete(&self, location: &Path) -> OSResult<()> {
+        self.0.delete(location).await
+    }
+
+    fn delete_stream<'a>(
+        &'a self,
+        locations: BoxStream<'a, OSResult<Path>>,
+    ) -> BoxStream<'a, OSResult<Path>> {
+        self.0.delete_stream(locations)
+    }
+
+    fn list(&self, prefix: Option<&Path>) -> BoxStream<'_, OSResult<ObjectMeta>> {
+        self.0.list(prefix)
+    }
+
+    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> OSResult<ListResult> {
+        self.0.list_with_delimiter(prefix).await
+    }
+
+    async fn copy(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.copy(from, to).await
+    }
+
+    async fn rename(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.rename(from, to).await
+    }
+
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.copy_if_not_exists(from, to).await
+    }
+}
+
+enum UploadState {
+    Pending,
+    CreatingUpload(BoxFuture<'static, OSResult<MultipartId>>),
+    InProgress {
+        multipart_id: Arc<MultipartId>,
+        part_idx: u16,
+        futures: FuturesUnordered<
+            BoxFuture<'static, std::result::Result<(u16, PartId), UploadPutError>>,
+        >,
+        part_ids: Vec<Option<PartId>>,
+    },
+    PuttingSingle(BoxFuture<'static, OSResult<()>>),
+    Completing(BoxFuture<'static, OSResult<()>>),
+    Done,
+}
+
+/// Start at 5MB.
+const INITIAL_UPLOAD_SIZE: usize = 1024 * 1024 * 5;
+
+struct Upload {
+    store: Arc<GoogleCloudStorage>,
+    path: Arc<Path>,
+    buffer: Vec<u8>,
+    state: UploadState,
+    connection_resets: u16,
+}
+
+impl Upload {
+    fn new(store: Arc<GoogleCloudStorage>, path: Path) -> OSResult<Self> {
+        Ok(Self {
+            store,
+            path: Arc::new(path),
+            buffer: Vec::with_capacity(INITIAL_UPLOAD_SIZE),
+            state: UploadState::Pending,
+            connection_resets: 0,
+        })
+    }
+
+    fn put_next_part(
+        path: &Arc<Path>,
+        store: &Arc<GoogleCloudStorage>,
+        buffer: &mut Vec<u8>,
+        part_idx: &mut u16,
```

Comment:
Very minor nit: This method always increases `part_idx` by 1 and clears `buffer`.  Maybe it would be better to just accept `Vec<u8>` and `u16` (instead of `&mut`) and let the caller do these things?  Normally I think of passing in `&mut` if it's going to be used recursively, or might be updated, or could be updated multiple times, etc.

Or...maybe this method can take in a `&mut self`?

Like I said, very minor, probably not worth worrying over.

Reply:
Okay. This is not separated into two methods: `next_part_buffer()` and `put_part()`.

Reason it doesn't take `&mut self` is we have values at the `Self` level as well as within the `UploadState::InProgress { .. }` struct variant. There's probably a nicer way to refactor this, but I don't want to get to aggressive, since this code is annoying to test at the moment.